recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,259 @@
+
+``` diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...  
+    }
+    providers: {
+      ...
+    }
+    recipeConfig: {
+      terraform: {
+        authentication: {
+          ...
+        }
+        providers: {
+          azurerm: [
+            {
+              subscriptionid: 1234
+              tenant_id: '745fg88bf-86f1-41af'
++              secrets: { // Individual Secrets from SecretStore
++                my_secret_1: {
++                  source: secretStoreConfig.id
++                  key: 'secret.one'
++                }
++                my_secret_2: {
++                  source: secretStoreConfig.id
++                  key: 'secret.two'
++                }
++              }
+            }, {
+              subscriptionid: 1234
+              tenant_id: '745fg88bf-86f1-41af'
+              alias: 'az-paymentservice'
+            } ]
+          gcp: [
+            {
+              project: 1234
+              regions: [ 'us-east1', 'us-west1' ]
+            }
+          ]
+          oraclepass: [
+            {
+              database_endpoint: '...'
+              java_endpoint: '...'
+              mysql_endpoint: '...'
+            }
+          ]
+        }
+      }
+      env: {
+        MY_ENV_VAR_1: 'my_value'
+      }
++      envSecrets: { // Individual Secrets from SecretStore
++        MY_ENV_VAR_2: {
++          source: secretStoreConfig.id
++          key: 'envsecret.one'
++        }
++        MY_ENV_VAR_3: {
++          source: secretStoreConfig.id
++          key: 'envsecret.two'
++        }
++     }
+    }
+    recipes: {
+    }
+  }
+}
+
+```
+
+**Sample Output:**
+n/a
+
+**Sample Recipe Contract:**
+n/a
+
+## Design
+
+### High Level Design
+
+The design for handling secrets for Private Terraform modules established the basis for data flow between Engine and Driver components in the codebase. For Terraform Providers, we will continue to maintain the flow where secrets are retrieved in the Engine and passed into the Driver in the Execute() call. 
+The following diagram describes the updated data flow. The calls in green highlight updates to existing code.
+
+
+![alt text](2024-04-terraform-provider-secrets.png)
+
+
+### Detailed Design
+
+The FindSecretIds function in the DriverWithSecrets interface is updated to return a struct of RecipeSecretIds, which will hold secret Ids for a Terraform module as before and in addition, also populate secret Ids for provider configuration and environment variables, iterating through the input environment recipe configuration. 
+
+The LoadSecrets method in the SecretLoader interface will also be updated to LoadRecipeSecrets and return struct RecipeSecrets which will contain secrets for module, providers and environment variables.
```

Comment:
Do other drivers have concepts like `module`, `providers`, `env-vars`?

Reply:
updated functions in design doc so we are not mirroring these constructs when handling secret data